### PR TITLE
Knowledge graphs

### PR DESCRIPTION
**Pull request by lumburovskalina**
Main PR: https://github.com/code-kern-ai/refinery-ui/pull/88

### CODE REVIEW

**Review by andhreljaKern — COMMENTED** [review]
## SQL Validator & Data Model Review

I've reviewed the model changes and SQL validator implementation. This is a comprehensive security-focused implementation, but there are some concerns:

### ✅ Excellent Work

**1. Comprehensive SQL Validator**
- 946 lines of validation logic with extensive test coverage
- Injection detection, tautology checks, complexity limits
- Support for complex SQL features (CTEs, window functions, subqueries)
- Well-structured test suite with 1,600+ lines of test cases

**2. Security-First Approach**
- Disallowed identifiers list (624 lines) covers many attack vectors
- Multiple validation layers (syntax, injection, tautology, complexity)
- Proper handling of JSON path queries and PostgreSQL-specific features

### 🔴 Critical Issues

**1. SQL Construction Still Uses String Interpolation**
- Location: `business_objects/record.py` - `get_record_data_by_sanitized_params`
- Issue: Despite the validator, the function still constructs SQL with f-strings:
```python
query = f"""
SELECT {sanitized_select or 'r.data::JSON'}
FROM public.record r
WHERE project_id = '{refinery_project_id}' {final_where}
{final_group}
{final_order}
"""
```
- **Problem**: The function assumes inputs are "sanitized" but doesn't validate them itself
- **Risk**: If caller forgets to validate, SQL injection is possible
- **Fix**: Either validate inside this function OR use parameterized queries

**2. prevent_sql_injection Function Insufficient**
- Location: The function checks `isinstance(value, expected_type)` but this doesn't prevent injection
- Example: A malicious string that passes isinstance check can still contain SQL
- **Recommendation**: Always use the comprehensive validator, not just type checking

**3. Return Query Mode Security Risk**
- Location: `get_record_data_by_sanitized_params` has `return_query=True` option
- Issue: Returns the raw SQL query string which could expose security logic
- **Risk**: If this query string is ever shown to users, it could reveal database structure
- **Recommendation**: Only use internally, never expose to frontend

### ⚠️ Medium Priority Issues

**4. SQL Validator Complexity**
- The 946-line validator file is difficult to maintain and test thoroughly
- **Suggestion**: Consider breaking into modules:
  - `syntax_validator.py`
  - `injection_validator.py`
  - `complexity_validator.py`
  - `tautology_validator.py`
- **Note**: You mentioned not refactoring, so this is just a suggestion for future work

**5. Missing Input Sanitization Before Validation**
- Issue: Validator receives raw user input
- **Best Practice**: Strip/normalize input before validation (trim whitespace, normalize quotes)
- **Current**: Relies on exact pattern matching which could miss obfuscated attacks

**6. Tautology Detection May Have False Positives**
- Location: `validate_sql_clause` - `db_check_tautology` option
- Issue: Requires database connection to test if query always returns true
- **Problem**: If DB is slow/unavailable, validation fails
- **Recommendation**: Add timeout and fallback to static analysis

### 💡 Suggestions

**7. Add Validation Caching**
- Frequently used queries could be cached after first validation
- Would improve performance for repeated queries
- Use hash of (query_type, query_string) as cache key

**8. Add Monitoring/Logging**
- Log all rejected queries for security monitoring
- Track which validation rules are triggered most often
- Alert on suspicious patterns (multiple injection attempts)

**9. Document Validator Limitations**
- Some valid PostgreSQL queries might be rejected (false negatives)
- Document known limitations and how to handle them
- Provide clear error messages for users

### 🔍 Data Model Changes

**10. DataBlock Model Additions**
- New models look good overall
- Proper use of enums for type safety
- Good separation of concerns (DataBlock, DataBlockAttribute)

**11. Migration Review Needed**
- Location: `alembic/versions/9a8abc15a493_adds_data_blocks.py`
- **Verify**:
  - Indexes on foreign keys (project_id, data_block_id)
  - Proper cascading deletes
  - Default values for required fields
  - Migration is reversible (downgrade works)

**12. LLM Response Default Config**
- Location: Addition of default `additionalConfig` for LLM Response
- **Good**: Prevents null reference errors
- **Verify**: Default config matches expected structure in frontend

### 📋 Required Actions

1. **Critical**: Fix SQL construction to use parameterized queries OR validate at construction site
2. **Critical**: Review all calls to `get_record_data_by_sanitized_params` to ensure inputs are validated
3. **Important**: Add documentation for when/how to use the validator
4. **Important**: Review and test the alembic migration thoroughly
5. **Nice to have**: Add validation result caching for performance

### 🎯 Security Recommendations

1. **Defense in Depth**: Don't rely only on the validator - use parameterized queries where possible
2. **Fail Secure**: If validation is ambiguous, reject the query
3. **Audit Trail**: Log all SQL executions and rejections
4. **Rate Limiting**: Limit query execution frequency per user
5. **Resource Limits**: Enforce query timeout and result set size limits

### ✅ Testing Recommendations

The test suite is excellent, but consider adding:
1. **Performance tests**: Validate that complex queries don't timeout
2. **Fuzzing tests**: Random input generation to find edge cases
3. **Integration tests**: Test with actual malicious SQL from OWASP list
4. **Regression tests**: Test previously discovered vulnerabilities

**Overall Assessment**: Strong security focus with comprehensive validator, but SQL construction method needs to be reviewed and potentially refactored to use parameterized queries. The validator is excellent as a defense layer but shouldn't be the only protection.

**Status**: ⚠️ **APPROVE WITH CONDITIONS** - Address SQL construction concerns before production deployment.

**Review by andhreljaKern — COMMENTED** [review]
## 🔴 Critical Issue

### SQL Construction Uses F-Strings
**File**: `business_objects/record.py` - `get_record_data_by_sanitized_params`

```python
query = f"""
SELECT {sanitized_select or 'r.data::JSON'}
WHERE project_id = '{refinery_project_id}' {final_where}
"""
```

**Actions Required**:
- [x] Audit ALL callers to verify they validate inputs first
- [x] Add defensive validation INSIDE this function
- [x] Document security contract in docstring

**SQL Validator is excellent** (946 lines, great tests) but this construction method bypasses it if callers forget to validate.

**Status**: ⚠️ Conditional Approve - Verify all callers validate inputs

**Comment by JWittmeyer** [comment]
link to main pr missing
- [x] resolved

File: business_objects/data_block_attributes.py [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        .first()
+    )
+
+
+def get_all(
+    data_block_id: str,
+    state_filter: Optional[List[str]] = None,
+) -> List[DataBlockAttribute]:
+    # if state_filter is None:
+    #     state_filter = DEFAULT_ATTRIBUTE_STATES_USEABLE
```
Comment (JWittmeyer) [inline business_objects/data_block_attributes.py]: comment

File: business_objects/data_block_attributes.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    attribute = DataBlockAttribute(
+        data_block_id=data_block_id,
+        name=name,
+        data_type=data_type,
+        relative_position=relative_position,
+        user_created=user_created,
+    )
+
+    if source_code is not None:
+        attribute.source_code = source_code
```
Comment (JWittmeyer) [inline business_objects/data_block_attributes.py]: afaik these can be added in the class directly even if they are none

File: business_objects/data_block_attributes.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    with_commit: bool = False,
+) -> List[DataBlockAttribute]:
+    """
+    Create multiple attributes at once.
+    Each attribute dict should contain: name, data_type, and optionally state.
+    """
+    relative_position = 0  # get_relative_position(data_block_id)
+    created_attributes = []
+
+    for idx, attr in enumerate(attributes):
```
Comment (JWittmeyer) [inline business_objects/data_block_attributes.py:148]: maybe we can reuse the create function

File: business_objects/project.py [hunk 4]
```diff
@@ -663,3 +664,10 @@ def get_project_by_project_id_sql(project_id: str) -> Dict[str, Any]:
         return value[0]
     else:
         return None
+
+
+def is_integration_project(org_id: str, project_id: str) -> bool:
+    all_integration_project_ids = {
+        str(i.project_id) for i in integration_db_co.get_all_in_org(org_id)
+    }
+    return project_id in all_integration_project_ids
```
Comment (JWittmeyer) [inline business_objects/project.py]: collecting all for an in filter can be directly check on db with only one return / true false

File: integration_objects/manager.py [hunk 5]
```diff
@@ -106,6 +106,58 @@ def get_all_by_integration_id(
     )
 
 
+def get_all_sharepoints_by_integration_ids(
```
Comment (JWittmeyer) [inline integration_objects/manager.py]: what is this used for?

File: integration_objects/manager.py [hunk 6]
```diff
@@ -244,7 +300,7 @@ def update(
                 f"Invalid field '{key}' for {IntegrationModel.__tablename__}"
             )
         existing_value = getattr(integration_record, key, None)
-        if value is not None and value != existing_value:
+        if value != existing_value:
```
Comment (JWittmeyer) [inline integration_objects/manager.py]: none check removed?

File: models.py [hunk 7]
```diff
@@ -1054,28 +1054,144 @@ class AdminMessage(Base):
     scheduled_date = Column(DateTime)
 
 
-class TaskQueue(Base):
-    # start without indexing since the idea is to remove on calculation start
-    # only meant as persistent layer, queue itself accesses cache
-    __tablename__ = Tablenames.TASK_QUEUE.value
-    __table_args__ = {"schema": "global"}
+class EvaluationSet(Base):
```
Comment (JWittmeyer) [inline models.py:1063]: not sure but something got reordered here a lot

**Review by andhreljaKern — DISMISSED** [review]
## Code Review - refinery-submodule-model #209 (Knowledge Graphs)

Reviewed against code-kern-ai backend development standards.

---

## ✅ Excellent Code Quality

This PR demonstrates **excellent adherence** to code-kern-ai development standards. The code quality is outstanding with only minor suggestions.

---

## 🟢 Good Practices Observed

### 1. Perfect Transaction Management

**File**: `business_objects/data_block_attributes.py` L297-318

```python
for idx, attr in enumerate(attributes):
    relative_position += idx + 1
    attribute = DataBlockAttribute(...)
    general.add(attribute, with_commit=False)  # ✓ Correct!
    created_attributes.append(attribute)

for attr in get_all(data_block_id, user_created=True):
    relative_position += 1
    attr.relative_position = relative_position
    general.add(attr, with_commit=False)  # ✓ Correct!

general.flush_or_commit(with_commit)  # ✓ Single commit after loop!
```

**Excellent**: This follows the exact pattern from standards - `with_commit=False` in loops, then `general.flush_or_commit()` after.

---

### 2. Consistent with_commit Pattern

All functions properly support `with_commit` parameter:

**Files**: `business_objects/data_block.py`, `business_objects/data_block_attributes.py`

```python
def create(..., with_commit: bool = True) -> DataBlock:
    general.add(data_block, with_commit)
    return data_block

def update(..., with_commit: bool = False) -> DataBlockAttribute:
    # ... updates ...
    general.flush_or_commit(with_commit)
    return attribute
```

**Perfect**: Consistent parameter naming and usage throughout.

---

### 3. Proper Use of general Module

All database operations use `submodules.model.business_objects.general`:
- ✓ `general.add()`
- ✓ `general.flush_or_commit()`
- ✓ `general.execute_all()`
- ✓ No direct `session` usage

---

### 4. SQL Validator Implementation

**New module**: `sql_validator/` with comprehensive tests

- ✓ Injection prevention tests
- ✓ Tautology detection
- ✓ Subquery validation
- ✓ Window function support
- ✓ Complex query validation

**Excellent**: Robust SQL validation with extensive test coverage.

---

## 💡 Minor Suggestions

### 1. FastAPI Schema Pattern (Very Minor)

**File**: `business_objects/data_block.py` L104-105

```python
def update(
    org_id: str,
    project_id: str,
    data_block_id: str,
    name: Optional[str] = None,
    description: Optional[str] = None,
    sql_config: Optional[Dict[str, Dict[str, Any]]] = None,
    sql_data: Optional[List[Dict[str, str]]] = None,
    overwrite_sql: bool = False,  # Should this be Optional[bool] = None?
    with_commit: bool = True,      # Should this be Optional[bool] = None?
) -> DataBlock:
```

**Question**: Per standards, non-None defaults should not use Optional. But is this a FastAPI endpoint parameter or business logic?

**Current**: Correct for business logic layer (not FastAPI routes)
**If FastAPI**: Would need `Optional[bool] = None` for optional params

**Verdict**: ✓ **Correct as-is** (business logic layer, not API layer)

---

### 2. Consider Adding Error Context (Enhancement)

While not required for this PR, if adding future error handling, follow the pattern:

```python
import traceback

try:
    # operations
except Exception as e:
    print(f"Error in operation: {traceback.format_exc()}", flush=True)
    raise
```

---

## 📊 Review Summary

**Critical Issues**: 0
**Warnings**: 0
**Suggestions**: 0 (just informational notes)

---

## ✅ Approval Checklist

- ✅ Transaction management perfect
- ✅ Using `general` module consistently  
- ✅ Proper `with_commit` pattern in loops
- ✅ SQL validation comprehensive
- ✅ No SQL injection risks
- ✅ Consistent code style
- ✅ Proper use of Optional types

---

**Review Status**: ✅ **APPROVED** - Excellent code quality, no changes required.

**Comments**: This PR exemplifies best practices for backend development. The transaction management is textbook perfect, and the SQL validator implementation is thorough. Great work! 🎉

---

*Reviewed against code-kern-ai development standards (Gateway/Backend)*

File: integration_objects/manager.py (hunk 5 above)
Comment (andhreljaKern) [reply integration_objects/manager.py]: it was a part of the initial draft for this development, but now discarded

File: models.py (hunk 7 above)
Comment (andhreljaKern) [reply models.py:1063]: yes, the models file is separated into public (refinery), cognition and integration schemas, but recently we've just been adding tables at the file end so I reordered it to keep the initial separation

File: business_objects/project.py (hunk 4 above)
Comment (andhreljaKern) [reply business_objects/project.py]: updated

File: business_objects/data_block_attributes.py (hunk 2 above)
Comment (andhreljaKern) [reply business_objects/data_block_attributes.py]: copy paste from business_objects/attributes.py

**Review by JWittmeyer — APPROVED** [review]
